File: worker/src/telegram.js
/** Telegram alert helper. */

function escapeMd(s) {
  return String(s).replace(/[_*[\]()~`>#+\-=|{}.!\\]/g, "\\$&");
}

export async function telegramAlert(env, markdown) {
  if (!env.TELEGRAM_BOT_TOKEN || !env.TELEGRAM_CHAT_ID) return;
  try {
    await fetch(`https://api.telegram.org/bot${env.TELEGRAM_BOT_TOKEN.trim()}/sendMessage`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        chat_id: env.TELEGRAM_CHAT_ID.trim(),
        text: markdown,
        parse_mode: "MarkdownV2",
      }),
    });
  } catch (err) {
    console.error("Telegram alert failed:", err);
  }
}

export { escapeMd };

File: src/telegram.py
"""Minimal Telegram Bot API wrapper — just sendMessage with MarkdownV2."""
from __future__ import annotations

import os

import requests


def send(text: str, *, dry_run: bool = False) -> None:
    if dry_run:
        print("=" * 60)
        print(text)
        print("=" * 60)
        return

    token = os.environ["TELEGRAM_BOT_TOKEN"]
    chat_id = os.environ["TELEGRAM_CHAT_ID"]
    url = f"https://api.telegram.org/bot{token}/sendMessage"
    resp = requests.post(
        url,
        json={
            "chat_id": chat_id,
            "text": text,
            "parse_mode": "MarkdownV2",
            "disable_web_page_preview": True,
        },
        timeout=30,
    )
    if not resp.ok:
        # Telegram returns a useful description in the body — surface it.
        raise RuntimeError(f"Telegram send failed: {resp.status_code} {resp.text}")
